Handle containerd logs

achntrl opened the pull request:
### What does this PR do?

It Parses containerd logs to strip the header. In order to have agnostic tailers, we move the parsing logic to the decoder. The scheduler is responsible to add a parser to the LogSource

### Motivation

Containerd generate log files with a header 

```
2018-09-20T11:54:11.753589172Z stdout F {"a":"b","c":"d"}
```

In that case, we are able to remap the message in pipeline but we can't automatically parse JSON. We have to strip the header in order to enable automatic log parsing


### Additional Notes

TODO: 
- [x] Handle the multiline case for Containerd
- [x] Initialize `LogSource` with IdentityParser
- [x] The docker tailer is not agnostic (it calls the docker parser). Find a better way of doing this
- [x] Add Reno

achntrl committed:
Move parsing logic to decoder/line_handler

achntrl committed:
Delete docker.go

It has been move to parser

achntrl committed:
Initialize decoder with empty Identity parser

achntrl committed:
Add docker parser in docker launcher

achntrl committed:
Move parsing logic to decoder

achntrl committed:
Add reno

iksaif commented on pkg/logs/parser/parser.go:
```diff
@@ -0,0 +1,268 @@
+func (p *IdentityParser) Unwrap(msg []byte) ([]byte, error) {
+	return msg, nil
+}
+
```
Would it make sense to move that to `parser/containerd.go` and `parser/docker.go` ?

iksaif commented on pkg/logs/parser/parser.go:
```diff
@@ -0,0 +1,268 @@
+	Timestamp string
+}
+
+// Parser parse containerd messages
```
It doesn't only parse containerd messages right ?

iksaif commented on pkg/logs/config/source.go:
```diff
@@ -57,3 +61,8 @@ func (s *LogSource) GetInputs() []string {
 }
+
+// AddParser add a parser
+func (s *LogSource) AddParser(pm parser.Parser) {
```
When I first saw `AddParser` I though this will add a parser to a stack of parsers. Maybe call it `SetParser` instead?

ajacquemot reviewed (changes requested):
Finish to review all the design of the code, it looks good so far but I think we can move around some stuff and and simplify a bit the `Parser` interface.
I did not read the `Parse` implementation for containerd yet, I'll have a look after your first round of changes.

achntrl committed:
Refactor parser package

achntrl committed:
Refactor k8s/launcher.go:getSource

achntrl commented:
@iksaif @ajacquemot Thank you for your feedback, I fixed all the things I agreed on

* I put an `Unwrap` method in the `Parser` to have a faster method for stripping the headers (this method will be called on every single line of log)
* I put the `SetParser` logic in the `launcher` because it already has the responsibility to start the right kind of `tailer`. Putting it in the `tailer` would force the `tailer` to know who created it and it won't be agnostic anymore. Putting it in the `scheduler` would be messy as it would have to handle all the cases, I think it's more natural to put it in the `launcher` as most of the dispatching logic naturally occurs with the launchers `select case` loop. Also, adding the `parser` close to the starting of the `tailer` avoid bugs where the `scheduler` would add a `Parser` and the `source` is consumed by the wrong `launcher`
* I don't think it makes sense to put the `parser` logic in the `message` package because it doesn't interact with messages. What happens is the tailer call the decoder that call the parser. The parser returns to the decoder that returns `Output`s to the tailer. then the tailer create a message and send it on its output channel.
- [ ] I have to look a bit more into error handling. I think what we want is to skip the message if it errors on parsing (that is the current behaviour)

achntrl committed:
Fix tests

achntrl committed:
Move message/status to its own package

achntrl committed:
Error handle parsing errors

iksaif commented on pkg/logs/config/source.go:
```diff
@@ -57,3 +61,10 @@ func (s *LogSource) GetInputs() []string {
+	s.lock.Lock()
+	s.Parser = parser
+	s.lock.Unlock()
+}
```
As explained IRL, it could be worth having a thread-safe `GetParser()`, but currently it would not change anything.

iksaif commented on pkg/logs/decoder/line_handler.go:
```diff
@@ -82,12 +87,22 @@ func (h *SingleLineHandler) process(line []byte) {
-		output := NewOutput(content, lineLen+1)
+		parsedMessage, err := h.parser.Parse(content)
+		if err != nil {
+			log.Warn(err)
```
would it be worth making `process` return an error ?

achntrl replied:
I don't think so because `process` runs in a goroutine and loops indefinitely
The communication with the tailer is done through a channel, so when there's an error, we don't push a message in the channel (before, the parsing logic was in the tailer and the tailer would skip the message on error)

iksaif reviewed (commented):
A few comments, nothing big. It mostly looks good to me but I'll let @ajacquemot make the call.

mnshdw commented on pkg/logs/parser/docker.go:
```diff
@@ -99,24 +106,24 @@ func getStatus(msg []byte) string {
-// that is in front of each message.
-func GetDockerMetadataLength(msg []byte) int {
+// getDockerMetadataLength returns the length of the 8 bytes header, timestamp, and space
+// that is in front of each config.
```
Not sure about this documentation change

mnshdw commented on pkg/logs/parser/parser.go:
```diff
@@ -0,0 +1,37 @@
+package parser
+
+// ParsedLine represents a containerd message
+type ParsedLine struct {
```
ParsedLine is *really* similar to Message, they're almost the same. Can we keep a single one ?

mnshdw commented on pkg/logs/parser/docker.go:
```diff
@@ -3,22 +3,15 @@
 // Copyright 2018 Datadog, Inc.
 
-package docker
+package parser
```
I'd rather keep docker in its own package and implement the parser there than the reverse, because I feel like the pattern docker.DoXYZ(...) is better than parser.DoDockerXYZ(...), but that is debatable. We may need to review how we organize our packages so we don't end up with a mess, happy to chat about it.

mnshdw replied:
cc @ajacquemot for input

ajacquemot replied:
looks nice when using the parser but I'd avoid having a docker package that only contains the parsing logic.

mnshdw commented on pkg/logs/parser/containerd.go:
```diff
@@ -0,0 +1,109 @@
+	endOfTimestampIdx := bytes.Index(msg, []byte{' '})
+	if endOfTimestampIdx == -1 {
+		// Nothing after the timestamp: ERROR
+		return ParsedLine{}, errors.New("can't parse containerd message")
```
I think you could add more detail to the error messages, eg: "can't parse containerd message, no whitespace found after timestamp"

mnshdw commented on pkg/logs/parser/containerd.go:
```diff
@@ -0,0 +1,109 @@
+	endOfLogStreamTypeIdx += endOfTimestampIdx + 1
+	severity := getContainerdSeverity(msg[endOfTimestampIdx+1 : endOfLogStreamTypeIdx])
+
+	endOfPartialFlagIdx := bytes.Index(msg[endOfLogStreamTypeIdx+1:], []byte{' '})
```
nit: Imho the code would easier to read if you simplified a bit the variable names, eg timestamp, type, partial.. There are no other variables so the fact that they are all offsets is rather clear.

mnshdw commented on pkg/logs/parser/parser.go:
```diff
@@ -0,0 +1,37 @@
+
+// Parse does nothing for NoopParser
+func (p *NoopParser) Parse(msg []byte) (ParsedLine, error) {
+	return ParsedLine{Content: msg, Severity: ""}, nil
```
Do we need `Severity: ""` ?

mnshdw commented on pkg/logs/config/source.go:
```diff
@@ -30,6 +33,7 @@ func NewLogSource(name string, config *LogsConfig) *LogSource {
 		inputs:   make(map[string]bool),
 		lock:     &sync.Mutex{},
 		Messages: NewMessages(),
+		Parser:   parser.NewNoopParser(),
```
I'm still not 100% sold on this.. I liked the fact that LogSource was a rather descriptive object and not tied to implementation details like how we parse logs. Have you considered other alternatives, like only setting a type and letting the line handler pick a parser (or none) depending on it ?

mnshdw reviewed (changes requested):
I did a first review ad left many nitpicks that are not that important, but I think there are issues that we should address before this goes live. Most importantly: we should not drop logs carelessly. Also, code-wise, I think we could maybe do better than fitting the parser in the log source. Finally, it would be cool to keep either ParsedLine or Message, not both as they are mostly the same and rather confusing.

achntrl committed:
Nits: improve naming in parser

achntrl committed:
Refactor parser to be reused

achntrl committed:
Move parser in their packages

achntrl committed:
Remove ParsedLine and Output

The parser now generate Messages that are enriched by the line handlers
and the tailers

achntrl committed:
Move status back to message package

achntrl committed:
Fix: Move Message to a reference for windowsevent

ajacquemot commented on pkg/logs/input/file/parser.go:
```diff
@@ -0,0 +1,109 @@
+// 2018-09-20T11:54:11.753589172Z stdout F This is my message
+func (p *parser) Parse(msg []byte) (*message.Message, error) {
+	// timestamp goes till first space
+	timestamp := bytes.Index(msg, []byte{' '})
```
Would be nice to mimic what's done in `docker/parser` https://github.com/DataDog/datadog-agent/blob/1e5ef8239681daf319cd834897dcf2ddaa380b45/pkg/logs/input/docker/parser.go#L35

ajacquemot reviewed (commented):
Left few comments and still few comments to address, we're getting very close.

mnshdw commented on pkg/logs/config/source.go:
```diff
@@ -9,16 +9,22 @@ import (
+	inputs       map[string]bool
+	lock         *sync.Mutex
+	Messages     *Messages
+	parserFormat string
```
I think we can avoid this altogether by moving the check to the container id in the tailer.

achntrl committed:
Refactor file/parser.go

achntrl committed:
Refactor message

Refactor message

Fix tests

achntrl committed:
Nits: remove useless variable

achntrl committed:
Refactor sourceType

ajacquemot commented on pkg/logs/config/source.go:
```diff
@@ -19,6 +19,10 @@ type LogSource struct {
+	// sourceType is the type of the source that we are tailing whereas Config.Type is the type of the tailer
+	// that reads log lines for this source. E.g, a sourceType == containerd and Config.Type == file means that
+	// the agent is tailing a file to read logs of a containerd container
+	sourceType string
```
nits: sourceType -> \`type\`

achntrl replied:
`type` is a reserved keyword, that's why I put `sourceType`

ajacquemot commented on pkg/logs/input/file/parser.go:
```diff
@@ -0,0 +1,82 @@
+// Example:
+// 2018-09-20T11:54:11.753589172Z stdout F This is my message
+func (p *parser) Parse(msg []byte) (*message.Message, error) {
+	parsedMsgChunks, err := parseMsg(msg)
```
nits: `components`

ajacquemot commented on pkg/logs/input/file/parser.go:
```diff
@@ -0,0 +1,82 @@
+	}
+	severity := getContainerdStatus(parsedMsgChunks[1])
+
+	parsedMsg := message.NewMessage()
```
nits, `msg`

achntrl replied:
`msg` is already the name of the function input

ajacquemot commented on pkg/logs/input/file/parser.go:
```diff
@@ -0,0 +1,82 @@
+	}
+	severity := getContainerdStatus(parsedMsgChunks[1])
+
+	parsedMsg := message.NewMessage()
```
timestamp is missing here

achntrl replied:
We don't need the timestamp here because the offset of the tailer if computer from RawDataLen

ajacquemot commented on pkg/logs/input/file/parser.go:
```diff
@@ -0,0 +1,82 @@
+
+// Unwrap remove the header of log lines of containerd
+func (p *parser) Unwrap(line []byte) ([]byte, error) {
+	parsedMsgChunks, err := parseMsg(line)
```
nits: `chunks` or `components`, prefer short variable names

ajacquemot commented on pkg/logs/input/file/parser.go:
```diff
@@ -0,0 +1,82 @@
+
+// getContainerdStatus returns the severity of the message based on the value of the
+// STREAM_TYPE field in the header
+func getContainerdStatus(logStream []byte) string {
```
nits: `logStream` -> `stramType`

ajacquemot commented on pkg/logs/input/file/parser.go:
```diff
@@ -0,0 +1,82 @@
+	case stderr:
+		return message.StatusError
+	default:
+		return ""
```
always return `info` by default

ajacquemot replied:
worth adding a comment too

ajacquemot commented on pkg/logs/input/file/parser.go:
```diff
@@ -0,0 +1,82 @@
+	}
+}
+
+func parseMsg(msg []byte) ([][]byte, error) {
```
nits `parse` or `parseMessage`, I vote for `parse`

ajacquemot commented on pkg/logs/input/file/parser.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+func parseMsg(msg []byte) ([][]byte, error) {
+	parsedMsgChunks := bytes.SplitN(msg, []byte{' '}, 4)
```
ditto `chunks` or `components`

ajacquemot commented on pkg/logs/input/file/parser.go:
```diff
@@ -0,0 +1,82 @@
+
+	// Empty message
+	if len(parsedMsgChunks) == 3 {
+		parsedMsgChunks = append(parsedMsgChunks, []byte(nil))
```
I think we should drop this message, this what we do with `docker`

ajacquemot commented on pkg/logs/input/journald/tailer.go:
```diff
@@ -174,12 +174,12 @@ func (t *Tailer) shouldDrop(entry *sdjournal.JournalEntry) bool {
-		t.getStatus(entry),
-	)
+func (t *Tailer) toMessage(entry *sdjournal.JournalEntry) *message.Message {
+	msg := message.NewMessage()
```
I think we should add a new constructor for message:
```
NewMessage(content, origin, status)
```

ajacquemot commented on pkg/logs/input/kubernetes/launcher.go:
```diff
@@ -129,7 +129,11 @@ func (l *Launcher) addSource(service *service.Service) {
 	}
 
-	l.sourcesByContainer[service.GetEntityID()] = source
+	if svc.Type == service.Containerd {
```
why do you need to check the service type ? I think you could just call:
```
source.SetSourceType(config.ContainerdType)
```

achntrl replied:
Because the service type could be docker, and we don't want to apply any parser in that case (because the logs are already json formatted)

achntrl commented on pkg/logs/decoder/line_handler.go:
```diff
@@ -82,12 +87,22 @@ func (h *SingleLineHandler) process(line []byte) {
-		output := NewOutput(content, lineLen+1)
+		parsedMessage, err := h.parser.Parse(content)
+		if err != nil {
+			log.Warn(err)
```
This is the old behaviour, I created another ticket to address this (this PR is _starting_ to get big)

ajacquemot commented on pkg/logs/input/windowsevent/tailer.go:
```diff
@@ -61,21 +61,21 @@ func (t *Tailer) Identifier() string {
-		message.NewOrigin(t.source),
-		message.StatusInfo,
-	), nil
+	msg := message.NewMessage()
```
same here, need to use a new constructor to make this cleaner, shorter

ajacquemot commented on pkg/logs/message/message.go:
```diff
@@ -6,47 +6,27 @@
-	origin  *Origin
-	status  string
+// NewMessage creates an empty new message
+func NewMessage() *Message {
```
I think you need two constructors:
```
func NewMessage(content []byte, origin *Origin, status string) *Message
func NewEmptyMessage()
```

ajacquemot commented on releasenotes/notes/handle-containerd-text-logs-bf47bdfbca65eddc.yaml:
```diff
@@ -0,0 +1,5 @@
+---
+enhancements:
+  - |
+    Strip header for containerd text logs and add it to the metadata of
```
I don't think we should add a release note for this feature.

ajacquemot reviewed (commented):
Mostly nits, great work, thanks for working on this, one last sprint and we are over. Still need to look at the tests.

achntrl committed:
Address nits

achntrl committed:
Revert "Add reno"

This reverts commit 78965173030d46de2ea71e33f0f82cdd26f26203.

achntrl committed:
Set source type in all cases

achntrl committed:
nits: fix naming issues on message

ajacquemot reviewed (approved):
LGTM, let's merge it ASAP
